UOW sample using the pipeline

andreasohlund opened the pull request:
Needs a review and also a discussion about how to dispose the session (now leaning on the container)

Related to https://github.com/Particular/NServiceBus/issues/4281

andreasohlund commented:
@DavidBoike @SimonCropp @timbussmann can you give this a first pass

timbussmann commented on samples/pipeline/unit-of-work/Core_6/Endpoint/Program.cs:
```diff
@@ -0,0 +1,137 @@
+    }
+}
+
+class MyUowBehavior : Behavior<IIncomingPhysicalMessageContext>
```
isn't this scenario something which you can already achieve with the current UoW interface? Do we need to show how to access data from the context?

andreasohlund replied:
Duh, indeed, I'll update to make use of header/message to create the session

DavidBoike replied:
Shouldn't the behavior be stateless and not use DI? Then within the Invoke method, use context.Builder to build the UoW-scoped session.

andreasohlund commented:
@timbussmann pushed https://github.com/Particular/docs.particular.net/pull/2228/commits/25efc81407aebd4eaea9fc60380f5a9fff28fd71 that makes use of headers when opening the uow.

The open methods looks kind of funny, do you have better ideas?

DavidBoike commented on samples/pipeline/unit-of-work/Core_6/Endpoint/Program.cs:
```diff
@@ -0,0 +1,154 @@
+    }
+}
+
+class SessionFactory
```
Not using this currently. Probably should register this as a SingleInstance. Then the registration for session has a choice whether to use the builder to build the SessionFactory, or use it directly in the closure. Not sure how to illustrate that or which approach is better?

andreasohlund replied:
I'll remove for now

andreasohlund commented:
@danielmarbach how do we handle `IDisposable` in this context? (asp. given how async makes it much trickier)

andreasohlund commented:
@DavidBoike addressed your comment. Still don't like the `Open` hack but I don't see another way to connect the message/headers with the session for the given pipeline instance?

DavidBoike commented:
A lot is balled up in what constitutes multi-tenancy right? Is multi-tenancy a SiteId on all queries? Or does it mean a different database? If the latter, and you were using NHibernate, you'd probably have a dictionary of session factories right?

danielmarbach commented on samples/pipeline/unit-of-work/Core_6/Endpoint/Program.cs:
```diff
@@ -0,0 +1,142 @@
+    {
+
+        var tennant = context.MessageHeaders["tennant"];
+        var session = context.Builder.Build<IMySession>();
```
Agree with @DavidBoike 

I'd say the only dependency it should have is a dependency to a session factory which is a singleton. Then it explicitly creates a sessions and adds it to the extension part of the context.

danielmarbach commented on samples/pipeline/unit-of-work/Core_6/Endpoint/Program.cs:
```diff
@@ -0,0 +1,142 @@
+{
+    readonly IMySession session;
+
+    public MyMessageHandler1(IMySession session)
```
Remove the ctor injection and access it via custom extension method over the message handler context

andreasohlund replied:
I added it on purpose since the user asked for it but I guess in v6 using the context is the way we recommend?

danielmarbach replied:
Well who owns the state? The handler or the pipeline execution? In my opinion it is the pipeline execution. Therefore the state should be exposed only via context.

andreasohlund commented:
Pushed an update that exposes the session on the context instead

andreasohlund commented:
@SimonCropp are you happy with the general direction of this? (I'll complete the .md and make it ready to pull in)

andreasohlund commented:
Added sample description. This one is now ready for review

SimonCropp commented on samples/pipeline/unit-of-work/Core_6/Endpoint/MySession.cs:
```diff
@@ -0,0 +1,40 @@
+    public Task Commit()
+    {
+        var entitiesStored = string.Join(",", entities);
+        Console.Out.WriteLine($"{entitiesStored} stored in tennant database: {tennant}DB by session {GetHashCode()}");
```
write to a logger

andreasohlund commented on samples/pipeline/unit-of-work/Core_6/Endpoint/Program.cs:
```diff
@@ -0,0 +1,49 @@
+        endpointConfiguration.SendFailedMessagesTo("error");
+
+        #region configuration
+        endpointConfiguration.RegisterComponents(r => r.ConfigureComponent(b => new MySessionProvider(), DependencyLifecycle.SingleInstance));
```
I could skip DI here and just pass the provider to the behavior?
